### ISSUE TITLE
Some problems with idea-reworked

### ISSUE THREAD

**Issue by larsgw** [citation-js/bibtex-parser-experiments#3]
One big problem is the question of what should be parsed when parsing syntax, and what should parsed when mapping to CSL. Consider also that Bib.TXT should be able to use the same mapping.

  - [ ] diacritics: when parsing syntax, as Bib.TXT and some BibTeX supports utf8
  - [ ] other known symbol commands and ligatures: when parsing syntax
  - [ ] except, fields tagged as `verbatim` or `url` in the specification should not have commands parsed, and then the *syntax* parser has to know about all the different fields.
  - [ ] although field data is available, URL escaping should be handled when mapping since Bib.TXT should probably have that behavior too
  - [ ] *name* field parsing: should probably be when mapping
  - [ ] list field parsing (splitting on " and "): should probably be when mapping
  - [ ] markup: should be done when mapping, as markup differs between formats
  - [ ] `crossref`: should be done when mapping

Less crucial things, maybe:

  - [ ] let people extend the constants (i.e., add commands, diacritics, ligatures)
  - [ ] (only) warn for mis-matched entry brackets

**Comment by larsgw** [comment]
I have a different mental model of this now, with two stages of parsing: one of entries, and one of values. _name_ (and _date_) field parsing is still at mapping for the moment, but `verbatim` and `uri` as well as markup and list fields are at stage two of parsing now, and implemented.

**Comment by retorquere** [comment]
What does the first stage do? It looks like it does more than tokenization, but assuming the first stage also does command -> unicode mapping, I'm not clear on how you brought verbatim parsing to the 2nd stage. What does `>` show up as in the 2nd stage?

**Comment by larsgw** [comment]
Yeah first part is full parsing of the file syntax, resolving `@string`s and concatenation, resulting in a data structure with parsed entries but verbatim values. The second part is parsing markup, commands etc. That also has to include name and date parsing because in the second part, all brackets are "resolved". Both parts include tokenization and further parsing.

All in all, not optimal so I might have to rethink it. But that would probably also involve rethinking the tokenizer (just category codes?) and as a result the whole grammar.

I also still have to make special cases for commands like `frac` and `vphantom`. And sentence casing I guess, though that still feels like something that would be part of a formatter, not a parser.

**Comment by retorquere** [comment]
But literal lists have the same issue as name lists. With "list fields" do you mean "literal lists"?

Why would `frac` require a special case? There's nothing special about `frac`, it's just a two-argument command.

**Comment by larsgw** [comment]
> But literal lists have the same issue as name lists. With "list fields" do you mean "literal lists"?

I mean any list field. Brackets are kept to the second stage, and while I did implement top-level " and " splitting, so to speak, I didn't realize that needed to be done for name fields as well until after posting the comment.

> Why would `frac` require a special case? There's nothing special about `frac`, it's just a two-argument command.

Because I currently don't have any argument commands yet, apart from formatting which I do special-case (and would like to keep doing so). So I'll probably use formatting commands, symbol commands and "other" commands implemented as functions in tandem.

> Sentence casing has to know about bracketing, and commands, because the brackets and the placement of commands within them influence the result in fairly intricate ways. If just the end-result is sentence-cased after all else is done, the results won't be right.

The brackets would be translated to `<span class="nocase">` (unless that's insufficient?), which would work for CSL. Things like BibTeX → RIS would not, so that's be a good reason. Can you give an example of the command thing? Do you mean command at the start of a word?

**Comment by retorquere** [comment]
I guess it could be corrected if enough information travels along? It gets complicated fairly quickly though; `{\u AB}` is equivalent to `ÅB`, but `{AB}` is equivalent  to `<span class="nocase">AB</span>`.

If you don't parse with arguments, how do you parse `\c{\u{E}}` or `\href`?

**Comment by retorquere** [comment]
Some of the command-and-brace interaction is described at https://retorque.re/zotero-better-bibtex/support/faq/#why-the-double-braces

**Comment by larsgw** [comment]
Right, diacritics are a special case too. `\href` used to be a symbol command producing no output and the argument would turn into regular case protection, but I realise I accidentally removed that just now.

I see the BibLaTeX manual also has documentation about their algorithm for sentence casing (page 253-255). However, not when it is applied exactly: it seems to be part of specific citation styles.

**Comment by retorquere** [comment]
Styles decide for themselves whether to apply sentence casing but given the complexities of sentence casing in bib(la)tex I can't imagine that the styles actually each implement the sentence casing themselves.

**Comment by larsgw** [comment]
No, BibLaTeX has a helper for that (`\MakeSentenceCase`). However, I don't know for which fields that is applied ("`title`-like fields" doesn't tell me that much), and each style can decide whether to apply the helper.

**Comment by retorquere** [comment]
Ah yes, that is true, but it remains that bib(la)tex expect the input to be title cased, and that CSL expects it sentence case.

**Comment by larsgw** [comment]
Going through the issues.

Checked

  - #3
  - #4 
  - af72788b0b78af0c8c96c38988d98a195bf35388
  - cae475f075a05d1c074485a061b08ed245170c7e
  - #7
  - #10
  - #11

TODO list:

  - [x] try the BBT test suite ([link](https://github.com/citation-js/bibtex-parser-experiments/issues/4#issuecomment-551292192))
  - [ ] annotations? ([link](https://github.com/citation-js/bibtex-parser-experiments/issues/4#issuecomment-690290319))
  - [ ] `\frac`, `\href` ([link](https://github.com/citation-js/bibtex-parser-experiments/issues/4#issuecomment-570730425))
  - [ ] `\emph` takes only the first character of the subsequent text (tokenizer thing?)
  - [ ] JabRef groups?
  - [ ] check for major simplification of tokenizer(s?)
  - [ ] crude recovery? ([link](https://github.com/citation-js/bibtex-parser-experiments/commit/af72788b0b78af0c8c96c38988d98a195bf35388#commitcomment-43359689))
  - [ ] `\vphantom` ([link](https://github.com/citation-js/bibtex-parser-experiments/issues/10#issuecomment-597363742))
  - [ ] figure out the deal with `\c{aa}` ([link](https://github.com/citation-js/bibtex-parser-experiments/issues/10#issuecomment-598495648))
  - [x] `\enquote`, maybe ([link](https://github.com/citation-js/bibtex-parser-experiments/issues/7#issuecomment-553463991))
